docs/lofi-hifi-backends.md

```diff
@@ -0,0 +1,279 @@
+- fsdp support.
+
+`ilab train phased` will be used in conjunction with `ilab checkpoint evaluate`. This command will run full scale inter-checkpoint evaluation on the given directory. An output dir will then hold then best checkpoint and all necessary data to run another `ilab train phased` command on.
+
+Much of the code for `ilab train phased` and `ilab train peft` will live in the CLI, and hopefully will be shared. However, plans to move much of this code to an external training library should eventually happen. This is most likely not in scope for the first version of this work.
+
+### Reasoning
+
+Plugging into hardware acceleration and multi-phase training is the logical next step for ilab. Ensuring we do this in a clean way that does not overload our current commands is also crucial. Many of the processes in the backend are confusing so we want to abstract some of the steps away from users while also giving them a reasonable amount of choice in configuring these new processes. However, maintaining the current laptop story is important to users without hardware access. Splitting these two paths into separate commands maintains the integrity of each.
+
+Not only maintaining the current laptop story but enhancing it. Qlora by itself is great, but having LoRA based training with the ability to run ZeRO phaes 1-3 or full FSDP accleration given the right hardware will improve the experience for people on 1 or 2 GPU systems (which are quite common).
+
+If we implement two training functions which diverge greatly, the design of the CLI will split and we will virtually be maintaining two different code stacks, so keeping the feature set aligned in every way except the usage of LoRA is a step in the right direction.
+
+
+### ilab model train integrated
+
+
+This command would take something like the following arguments
+
+
+--gpus=str , describes the amount of GPUs (of what is available) to use for this process. This comes in the form of: 0-1, 8, etc.
+--quantize=bool, enabled Qlora which basically loads the model in a quantized form so it can fit on a consumer GPU
+--accelerator=str (deepspeed, fsdp) describes the optimizer framework to use during training
+--learning-rate, int (?)
+--batch-len=int
+--input-dir=str, describes where the generated+mixed data lives to be trained on.
+--model-name=str, name of the model to be output
+--output-dir=str, where to put the model after training
+--num-epochs=int, the amount of epochs to run in this phase of training
+--device=str, the device type to load the model onto: cuda,rocm,cpu,mlx,mps
+++ many of the current train flags probably
+
+#### Implementation Specifics
+
+The Transformers library (which is currently what we use for training along with torch) has options for fsdp, deepspeed, and many of the related arguments currently in use elsewhere in the project. My idea here is to try and use these as best as we can in the CLI rather than make our own custom versions that we then have to maintain.
```

Comment:
```suggestion
The Transformers library (which is currently what we use for training along with torch) has options for fsdp, deepspeed, and many of the related arguments currently in use elsewhere in the project. The goal is to try and use these as best as we can in the CLI rather than make our own custom versions that we then have to maintain.
```